### DISCUSSION TITLE
函数参数的处理

### DISCUSSION THREAD

**Discussion by paprika0741** [arcsysu/SYsU-lang#42] [Q&A] [answered]
请问对于一个函数的参数列表，除了在llvm::Function::Create的时候将参数类型传入以及设置参数名，还需要做什么处理吗？需要加入BB的符号表中吗？函数主体如何获得参数呢？

**Comment by Yorkking** [comment, marked as answer]
需要加入到符号表，和局部变量类似。函数主体可以调用 (Function*) F->Args() 获得。比如：
```C++
for(auto& arg: F->Args()){

}
```

**Reply by paprika0741** [reply]
> 需要加入到符号表，和局部变量类似。函数主体可以调用 (Function*) F->Args() 获得。比如：
> 
> ```c++
> for(auto& arg: F->Args()){
> 
> }
> ```


请问是遍历函数参数列表，手动加入符号表吗？

**Reply by Yorkking** [reply]
假如你用的是 llvm 自带的符号表，你可以自己试一下要不要手动加入，看看不加入能否找到。

**Reply by paprika0741** [reply]
参考
```define i32 @test(i32 %p) #0 {
entry:
  %p.addr = alloca i32, align 4
  store i32 %p, i32* %p.addr, align 4
  %0 = load i32, i32* %p.addr, align 4
  %sub = sub nsw i32 %0, 1
  store i32 %sub, i32* %p.addr, align 4
  ret i32 0
}
```
对于参数p， 手动create一个局部变量p.addr， 将p对应的arg load进去，函数体内遇到变量p，使用p.addr去查找符号表

**Reply by Aurora-1412** [reply]
不带参数的函数没问题了，带参数的函数用for(auto &arg: TheFunction->args())仍然段错误怎么办orz

**Reply by Yorkking** [reply]
> 不带参数的函数没问题了，带参数的函数用for(auto &arg: TheFunction->args())仍然段错误怎么办orz

最好贴一下相关的代码看看？

**Comment by Aurora-1412** [comment]
最终找到是我创建ArrayRef<Type*&gt; params 作为CreateFunc的参数时，实际上params只被分配了空间，但并没有把值传进去，请问ArrayRef怎么用呢，我有一个函数，其返回值为ArrayRef<Type*&gt;类型。而这个函数内部，无论是
ArrayRef<Type*&gt; arr; arr.vec().push_back(xxx)
还是vector<string&gt; vec; vec.push_back(xxx); ArrayRef<Type*&gt; arr(vec);
这两种赋值方式都不能把值“真正”传给ArrayRef<Type*&gt;类型的变量




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "arcsysu/SYsU-lang"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年5月19日(星期四) 晚上8:29
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [arcsysu/SYsU-lang] 函数参数的处理 (Discussion #42)





  
不带参数的函数没问题了，带参数的函数用for(auto &amp;arg: TheFunction-&gt;args())仍然段错误怎么办orz
  
最好贴一下相关的代码看看？

**Comment by Aurora-1412** [comment]
明白了，vector可以自动转换为ArrayRef类型，不应该返回ArrayRef，不然引用会丢失




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "arcsysu/SYsU-lang"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年5月19日(星期四) 晚上8:29
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [arcsysu/SYsU-lang] 函数参数的处理 (Discussion #42)





  
不带参数的函数没问题了，带参数的函数用for(auto &amp;arg: TheFunction-&gt;args())仍然段错误怎么办orz
  
最好贴一下相关的代码看看？
